prober/src/main/java/google/registry/monitoring/blackbox/ProbingAction.java

```diff
@@ -0,0 +1,156 @@
+// limitations under the License.
+
+package google.registry.monitoring.blackbox;
+
+import static com.google.common.flogger.StackSize.SMALL;
+import static google.registry.monitoring.blackbox.Protocol.PROTOCOL_KEY;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.flogger.FluentLogger;
+import google.registry.monitoring.blackbox.handlers.ActionHandler;
+import io.netty.util.AttributeKey;
+import java.util.concurrent.TimeUnit;
+import org.joda.time.Duration;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.HashedWheelTimer;
+import io.netty.util.Timer;
+import java.util.concurrent.Callable;
+import javax.inject.Provider;
+
+/**
+ *Class that represents given action in sequence of probing
+ *
+ */
+
+
+public abstract class ProbingAction<O> implements Callable<ChannelFuture> {
+
+  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
+
+  /** Attribute Key that links channel to its {@link ProbingAction}*/
+  public final static AttributeKey<ProbingAction> PROBING_ACTION_KEY = AttributeKey.valueOf("PROBING_ACTION_KEY");
+
+
+  /** {@link ActionHandler} Associated with this {@link ProbingAction}*/
+  private ActionHandler actionHandler;
+
+
+  /**
+   * The requisite instance of {@link ActionHandler}, which is always the last {@link ChannelHandler} in the pipeline
+   */
+  public ActionHandler actionHandler() {
+    return actionHandler;
+  }
+
+
+  /** {@link Timer} that rate limits probing*/
+  private static final Timer timer = new HashedWheelTimer();
+
+
+  /** actual {@link Duration} of this delay*/
+  public abstract Duration delay();
+
+  /** message to send to server */
+  public abstract O outboundMessage();
+
+  /**
+   * @return {@link Channel} object that represents connection between prober client and server
+   */
+  public abstract Channel channel();
+
+  /**
+   * @return The {@link Protocol} instance that represents action to be tested by this part in sequences
+   */
+  public abstract Protocol protocol();
+
+  /**
+   *
+   * @return {@link Builder} that lets us build a new ProbingAction by customizing abstract methods
+   */
+  public abstract <O, B extends Builder<O, B, P>, P extends ProbingAction<O>> Builder<O, B, P> toBuilder();
+
+
+  /**
+   * The method that calls the {@link ActionHandler} to send a message down the channel pipeline
+   * @return future that denotes when the action has been successfully performed
+   */
+
+  @Override
+  public ChannelFuture call() {
+
+    //Sets Action Handler to appropriately the last channel in the pipeline
+    //Logs severe if the last channel in the pipeline is not
+    try {
+      this.actionHandler = (ActionHandler) this.channel().pipeline().last();
+    } catch(ClassCastException exception) {
```

Comment:
Needs a space between catch and the parenthesis.

Reply:
Done.